### PR TITLE
tls: reject certificate without either a subject CN or altnames

### PR DESCRIPTION
**Pull request by jpeach**
A server certificate must have either a subject CN or supported subject
alternative names. Certificates that have neither are a configuration
error, not an assertion, so check for this condition and throw an
exception rather than asserting.

Risk Level: Low
Testing: Unit tests
Docs Changes: N/A
Release Notes:
Fixed: #9182

### CODE REVIEW

File: source/extensions/transport_sockets/tls/context_impl.cc [hunk 1]
```diff
@@ -911,6 +911,12 @@ ServerContextImpl::ServerContextImpl(Stats::Scope& scope,
   if (config.tlsCertificates().empty()) {
     throw EnvoyException("Server TlsCertificates must have a certificate specified");
   }
+
+  // Compute the session context ID hash. We use all the certificate identities,
+  // since we should have a common ID for session resumption no matter what cert
+  // is used. We do this early because it can throw an EnvoyException.
+  const SessionContextID session_id = generateHashForSessionContexId(server_names);
```
Comment (lizan) [inline source/extensions/transport_sockets/tls/context_impl.cc]: As pointed in previous PR (after merge), do you mind fix this spell?

File: source/extensions/transport_sockets/tls/context_impl.cc [hunk 2]
```diff
@@ -986,33 +989,58 @@ ServerContextImpl::generateHashForSessionContexId(const std::vector<std::string>
...
     const int cn_index = X509_NAME_get_index_by_NID(cert_subject, NID_commonName, -1);
-    // It's possible that the certificate doesn't have CommonName, but has SANs.
     if (cn_index >= 0) {
       X509_NAME_ENTRY* cn_entry = X509_NAME_get_entry(cert_subject, cn_index);
       RELEASE_ASSERT(cn_entry != nullptr, "certificate subject CN should be present");
+
       ASN1_STRING* cn_asn1 = X509_NAME_ENTRY_get_data(cn_entry);
-      RELEASE_ASSERT(ASN1_STRING_length(cn_asn1) > 0, "");
+      if (ASN1_STRING_length(cn_asn1) <= 0) {
+        EVP_MD_CTX_cleanup(&md);
```
Comment (lizan) [inline source/extensions/transport_sockets/tls/context_impl.cc]: Just change type of `md` to `bssl::ScopedEVP_MD_CTX` so we don't have to worry about cleanup?

File: source/extensions/transport_sockets/tls/context_impl.cc (hunk 1 above)
Comment (jpeach) [reply source/extensions/transport_sockets/tls/context_impl.cc]: Yikes, I stared at this for 5 mins before I could see the problem :)

Comment (jpeach) [reply source/extensions/transport_sockets/tls/context_impl.cc]: Fixed.

File: source/extensions/transport_sockets/tls/context_impl.cc (hunk 2 above)
Comment (jpeach) [reply source/extensions/transport_sockets/tls/context_impl.cc]: Fixed.

**Pull request merged**